File: packages/automl/bench/bench-automl.mjs
/**
 * AutoML benchmark: wlearn JS on Friedman 1-3, moons, hastie.
 * Sizes: 500, 2000, 10000.
 *
 * Uses WASM model packages from sibling repos (../../../<lib>-wasm/).
 * Skips any model that can't be imported.
 */

import { autoFit } from '../src/auto-fit.js'
import { makeFriedman1, makeFriedman2, makeFriedman3, makeMoons, makeHastie, trainTestSplit } from './datasets.mjs'

// --- Dynamic model imports (graceful skip) ---

async function loadModels() {
  const models = {}
  const base = new URL('../../../../', import.meta.url).pathname

  const tryImport = async (name, path, cls) => {
    try {
      const mod = await import(path)
      models[name] = mod[cls]
      console.log(`  ${name}: loaded`)
    } catch (e) {
      console.log(`  ${name}: skipped (${e.message.split('\n')[0]})`)
    }
  }

  await tryImport('xgb', base + 'xgboost-wasm/src/index.js', 'XGBModel')
  await tryImport('linear', base + 'liblinear-wasm/src/index.js', 'LinearModel')
  await tryImport('svm', base + 'libsvm-wasm/src/index.js', 'SVMModel')
  await tryImport('knn', base + 'nanoflann-wasm/src/index.js', 'KNNModel')
  await tryImport('ebm', base + 'ebm-wasm/src/index.js', 'EBMModel')
  await tryImport('lgb', base + 'lightgbm-wasm/src/index.js', 'LGBModel')
  await tryImport('tsetlin', base + 'tsetlin-wasm/src/index.js', 'TsetlinModel')

  return models
}

function makeRegSpecs(modelMap) {
  const specs = []
  if (modelMap.xgb) specs.push({ name: 'xgb', cls: modelMap.xgb, params: { objective: 'reg:squarederror', numRound: 100 } })
  if (modelMap.linear) specs.push({ name: 'linear', cls: modelMap.linear, params: { solver: 11, C: 1.0 } })
  if (modelMap.svm) specs.push({ name: 'svm', cls: modelMap.svm, params: { svmType: 3, kernel: 2, C: 1.0, gamma: 0 } })
  if (modelMap.knn) specs.push({ name: 'knn', cls: modelMap.knn, params: { k: 5, task: 'regression' } })
  if (modelMap.ebm) specs.push({ name: 'ebm', cls: modelMap.ebm, params: { objective: 'regression' } })
  if (modelMap.lgb) specs.push({ name: 'lgb', cls: modelMap.lgb, params: { objective: 'regression', numRound: 100, verbosity: -1 } })
  return specs
}

function makeClsSpecs(modelMap) {
  const specs = []
  if (modelMap.xgb) specs.push({ name: 'xgb', cls: modelMap.xgb, params: { objective: 'multi:softprob', numRound: 100 } })
  if (modelMap.linear) specs.push({ name: 'linear', cls: modelMap.linear, params: { solver: 0, C: 1.0 } })
  if (modelMap.svm) specs.push({ name: 'svm', cls: modelMap.svm, params: { svmType: 0, kernel: 2, C: 1.0, gamma: 0, probability: 1 } })
  if (modelMap.knn) specs.push({ name: 'knn', cls: modelMap.knn, params: { k: 5, task: 'classification' } })
  if (modelMap.ebm) specs.push({ name: 'ebm', cls: modelMap.ebm, params: { objective: 'classification' } })
  if (modelMap.lgb) specs.push({ name: 'lgb', cls: modelMap.lgb, params: { objective: 'binary', numRound: 100, verbosity: -1 } })
  return specs
}

// --- Scoring ---

function r2Score(yTrue, yPred) {
  const n = yTrue.length
  let mean = 0
  for (let i = 0; i < n; i++) mean += yTrue[i]
  mean /= n
  let ssTot = 0, ssRes = 0
  for (let i = 0; i < n; i++) {
    ssTot += (yTrue[i] - mean) ** 2
    ssRes += (yTrue[i] - yPred[i]) ** 2
  }
  return ssTot === 0 ? 0 : 1 - ssRes / ssTot
}

function accuracy(yTrue, yPred) {
  let correct = 0
  for (let i = 0; i < yTrue.length; i++) {
    if (Math.round(yPred[i]) === yTrue[i]) correct++
  }
  return correct / yTrue.length
}

// --- Run single benchmark ---

async function runWlearn(specs, Xtrain, ytrain, Xtest, ytest, task, strategy) {
  if (specs.length === 0) return { score: null, time: 0 }

  const t0 = performance.now()
  try {
    const result = await autoFit(specs, Xtrain, ytrain, {
      strategy,
      cv: 5,
      seed: 42,
      task,
      nIter: 20,
    })
    const elapsed = (performance.now() - t0) / 1000

    if (!result.model) return { score: null, time: elapsed }

    const preds = result.model.predict(Xtest)
    const score = task === 'classification'
      ? accuracy(ytest, preds)
      : r2Score(ytest, preds)

    result.model.dispose()
    return { score, time: elapsed }
  } catch (e) {
    const elapsed = (performance.now() - t0) / 1000
    console.error(`  Error (${strategy}): ${e.message}`)
    return { score: null, time: elapsed }
  }
}

// --- Format helpers ---

function fmtScore(s) {
  if (s === null) return '  --  '
  return s.toFixed(4).padStart(6)
}

function fmtTime(t) {
  if (t === 0) return '  -- '
  if (t < 1) return `${(t * 1000).toFixed(0).padStart(4)}ms`
  return `${t.toFixed(1).padStart(5)}s`
}

// --- Main ---

async function main() {
  console.log()
  console.log('AutoML Benchmark: wlearn JS')
  console.log('Loading models...')

  const modelMap = await loadModels()
  const nModels = Object.keys(modelMap).length
  console.log(`${nModels} model(s) available`)
  console.log()

  if (nModels === 0) {
    console.log('No WASM model packages found. Link them with npm or adjust import paths.')
    process.exit(1)
  }

  const sizes = [500, 2000]
  const datasets = {
    friedman1: (n, seed) => ({ ...makeFriedman1(n, { seed }), task: 'regression' }),
    friedman2: (n, seed) => ({ ...makeFriedman2(n, { seed }), task: 'regression' }),
    friedman3: (n, seed) => ({ ...makeFriedman3(n, { seed }), task: 'regression' }),
    moons:     (n, seed) => ({ ...makeMoons(n, { seed }), task: 'classification' }),
    hastie:    (n, seed) => ({ ...makeHastie(n, { seed }), task: 'classification' }),
  }

  // Header
  console.log('| Dataset   |     n | wlearn-portfolio |  time | wlearn-random    |  time |')
  console.log('|-----------|------:|-----------------:|------:|-----------------:|------:|')

  for (const n of sizes) {
    for (const [name, gen] of Object.entries(datasets)) {
      const { X, y, task } = gen(n, 42)
      const { Xtrain, ytrain, Xtest, ytest } = trainTestSplit(X, y, { testSize: 0.2, seed: 42 })

      const specs = task === 'classification' ? makeClsSpecs(modelMap) : makeRegSpecs(modelMap)

      const portfolio = await runWlearn(specs, Xtrain, ytrain, Xtest, ytest, task, 'portfolio')
      const random = await runWlearn(specs, Xtrain, ytrain, Xtest, ytest, task, 'random')

      console.log(
        `| ${name.padEnd(9)} | ${String(n).padStart(5)} |` +
        ` ${fmtScore(portfolio.score).padStart(16)} | ${fmtTime(portfolio.time).padStart(5)} |` +
        ` ${fmtScore(random.score).padStart(16)} | ${fmtTime(random.time).padStart(5)} |`
      )
    }
  }

  console.log()
}

main().catch(e => { console.error(e); process.exit(1) })

File: py/bench/bench_automl.py
"""AutoML benchmark: wlearn vs TPOT vs AutoGluon.

Datasets: Friedman 1-3 (regression), moons + hastie (classification).
Sizes: 500, 2000, 10000 samples each.
"""

import time
import warnings

import numpy as np
from sklearn.datasets import make_friedman1, make_friedman2, make_friedman3, make_moons
from sklearn.model_selection import train_test_split

warnings.filterwarnings('ignore')

# --- wlearn imports ---

from wlearn.xgboost import XGBModel
from wlearn.liblinear import LinearModel
from wlearn.libsvm import SVMModel
from wlearn.nanoflann import KNNModel
from wlearn.ebm import EBMModel
from wlearn.automl import auto_fit

# --- Optional competitors ---

try:
    from tpot import TPOTRegressor, TPOTClassifier
    HAS_TPOT = True
except ImportError:
    HAS_TPOT = False

try:
    from autogluon.tabular import TabularPredictor
    import pandas as pd
    HAS_AUTOGLUON = True
except ImportError:
    HAS_AUTOGLUON = False


# ---------------------------------------------------------------------------
# Dataset generators
# ---------------------------------------------------------------------------

def make_hastie(n_samples, seed=42):
    """Hastie et al. ESL 10.2: y=1 if sum(x_i^2) > chi2_median(10)."""
    rng = np.random.RandomState(seed)
    X = rng.randn(n_samples, 10)
    chi2_median = 9.3418  # median of chi2(df=10)
    y = (np.sum(X ** 2, axis=1) > chi2_median).astype(int)
    return X, y


def generate_datasets(n_samples, seed=42):
    """Generate all benchmark datasets at given size."""
    datasets = {}

    # Regression
    X, y = make_friedman1(n_samples=n_samples, noise=1.0, random_state=seed)
    datasets['friedman1'] = ('regression', X, y)

    X, y = make_friedman2(n_samples=n_samples, noise=1.0, random_state=seed)
    datasets['friedman2'] = ('regression', X, y)

    X, y = make_friedman3(n_samples=n_samples, noise=1.0, random_state=seed)
    datasets['friedman3'] = ('regression', X, y)

    # Classification
    X, y = make_moons(n_samples=n_samples, noise=0.3, random_state=seed)
    datasets['moons'] = ('classification', X, y)

    X, y = make_hastie(n_samples=n_samples, seed=seed)
    datasets['hastie'] = ('classification', X, y)

    return datasets


# ---------------------------------------------------------------------------
# wlearn model specs
# ---------------------------------------------------------------------------

REG_MODELS = [
    ('xgb', XGBModel, {'objective': 'reg:squarederror', 'numRound': 100}),
    ('linear', LinearModel, {'solver': 11, 'C': 1.0}),
    ('svm', SVMModel, {'svmType': 3, 'kernel': 2, 'C': 1.0, 'gamma': 0}),
    ('knn', KNNModel, {'k': 5, 'task': 'regression'}),
    ('ebm', EBMModel, {'objective': 'regression'}),
]

CLS_MODELS = [
    ('xgb', XGBModel, {'objective': 'multi:softprob', 'numRound': 100}),
    ('linear', LinearModel, {'solver': 0, 'C': 1.0}),
    ('svm', SVMModel, {'svmType': 0, 'kernel': 2, 'C': 1.0, 'gamma': 0}),
    ('knn', KNNModel, {'k': 5, 'task': 'classification'}),
    ('ebm', EBMModel, {'objective': 'classification'}),
]


# ---------------------------------------------------------------------------
# Runner functions
# ---------------------------------------------------------------------------

def run_wlearn(models, X_train, y_train, X_test, y_test, task, strategy, scoring):
    """Run wlearn auto_fit and return (score, time_seconds)."""
    t0 = time.time()
    result = auto_fit(
        models, X_train, y_train,
        strategy=strategy, cv=5, seed=42, task=task, n_iter=20,
    )
    elapsed = time.time() - t0
    model = result['model']
    if model is None:
        return None, elapsed

    if task == 'classification':
        preds = model.predict(X_test)
        score = float(np.mean(preds == y_test))
    else:
        preds = model.predict(X_test)
        ss_res = np.sum((y_test - preds) ** 2)
        ss_tot = np.sum((y_test - np.mean(y_test)) ** 2)
        score = 1 - ss_res / ss_tot if ss_tot > 0 else 0.0

    model.dispose()
    return score, elapsed


def run_tpot(X_train, y_train, X_test, y_test, task):
    """Run TPOT and return (score, time_seconds)."""
    if not HAS_TPOT:
        return None, 0

    t0 = time.time()
    try:
        if task == 'classification':
            tpot = TPOTClassifier(
                max_time_mins=1, max_eval_time_mins=0.5,
                cv=5, n_jobs=1, verbose=0,
            )
        else:
            tpot = TPOTRegressor(
                max_time_mins=1, max_eval_time_mins=0.5,
                cv=5, n_jobs=1, verbose=0,
            )

        tpot.fit(X_train, y_train)
        preds = tpot.predict(X_test)
        elapsed = time.time() - t0

        if task == 'classification':
            score = float(np.mean(preds == y_test))
        else:
            ss_res = np.sum((y_test - preds) ** 2)
            ss_tot = np.sum((y_test - np.mean(y_test)) ** 2)
            score = 1 - ss_res / ss_tot if ss_tot > 0 else 0.0

        return score, elapsed
    except Exception as e:
        elapsed = time.time() - t0
        print(f'  TPOT error: {e}', flush=True)
        return None, elapsed


def run_autogluon(X_train, y_train, X_test, y_test, task):
    """Run AutoGluon and return (score, time_seconds)."""
    if not HAS_AUTOGLUON:
        return None, 0

    import tempfile
    import shutil

    # AutoGluon needs DataFrames
    cols = [f'f{i}' for i in range(X_train.shape[1])]
    train_df = pd.DataFrame(X_train, columns=cols)
    train_df['target'] = y_train
    test_df = pd.DataFrame(X_test, columns=cols)

    tmpdir = tempfile.mkdtemp()
    try:
        t0 = time.time()
        predictor = TabularPredictor(
            label='target',
            path=tmpdir,
            problem_type='binary' if task == 'classification' else 'regression',
            verbosity=0,
        )
        predictor.fit(train_df, time_limit=60, presets='medium_quality')
        preds = predictor.predict(test_df).values
        elapsed = time.time() - t0

        if task == 'classification':
            score = float(np.mean(preds == y_test))
        else:
            ss_res = np.sum((y_test - preds) ** 2)
            ss_tot = np.sum((y_test - np.mean(y_test)) ** 2)
            score = 1 - ss_res / ss_tot if ss_tot > 0 else 0.0
    except Exception as e:
        print(f'  AutoGluon error: {e}', flush=True)
        score = None
        elapsed = 0
    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)

    return score, elapsed


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def fmt_score(score):
    if score is None:
        return '  --  '
    return f'{score:6.4f}'


def fmt_time(t):
    if t == 0:
        return '  -- '
    if t < 1:
        return f'{t*1000:4.0f}ms'
    return f'{t:5.1f}s'


def main():
    sizes = [500, 2000, 10000]

    # Header
    systems = ['wlearn-portfolio', 'wlearn-random']
    if HAS_TPOT:
        systems.append('TPOT')
    if HAS_AUTOGLUON:
        systems.append('AutoGluon')

    header_parts = ['| Dataset   |     n |']
    sep_parts = ['|-----------|------:|']
    for s in systems:
        header_parts.append(f' {s:>16s} |  time |')
        sep_parts.append(f' {"-"*16}:|------:|')

    print(''.join(header_parts))
    print(''.join(sep_parts))

    for n in sizes:
        datasets = generate_datasets(n, seed=42)

        for name, (task, X, y) in datasets.items():
            X_train, X_test, y_train, y_test = train_test_split(
                X, y, test_size=0.2, random_state=42,
            )
            scoring = 'accuracy' if task == 'classification' else 'r2'
            models = CLS_MODELS if task == 'classification' else REG_MODELS

            row = f'| {name:9s} | {n:5d} |'

            # wlearn portfolio
            s, t = run_wlearn(models, X_train, y_train, X_test, y_test,
                              task, 'portfolio', scoring)
            row += f' {fmt_score(s):>16s} | {fmt_time(t):>5s} |'

            # wlearn random
            s, t = run_wlearn(models, X_train, y_train, X_test, y_test,
                              task, 'random', scoring)
            row += f' {fmt_score(s):>16s} | {fmt_time(t):>5s} |'

            # TPOT
            if HAS_TPOT:
                s, t = run_tpot(X_train, y_train, X_test, y_test, task)
                row += f' {fmt_score(s):>16s} | {fmt_time(t):>5s} |'

            # AutoGluon
            if HAS_AUTOGLUON:
                s, t = run_autogluon(X_train, y_train, X_test, y_test, task)
                row += f' {fmt_score(s):>16s} | {fmt_time(t):>5s} |'

            print(row)


if __name__ == '__main__':
    print()
    print('AutoML Benchmark: wlearn vs competitors')
    print(f'TPOT: {"available" if HAS_TPOT else "not installed"}')
    print(f'AutoGluon: {"available" if HAS_AUTOGLUON else "not installed"}')
    print()
    main()
    print()
